docs/ethsimulatev1-notes.md

```diff
@@ -0,0 +1,229 @@
+
+### Phantom blocks
+The multicall allows you to define on what block number your calls or transactions are being executed on. E.g, consider following call:
+```json
+{
+	"jsonrpc": "2.0",
+	"id": 1,
+	"method": "eth_simulateV1",
+	"params": [
+		{
+			"blockStateCalls": [
+				{
+					"blockOverrides": {
+						"number": "0x64"
+					},
+				},
+				{
+					"blockOverrides": {
+						"number": "0xc8"
+					},
+				}
+			]
+		},
+		"0xa"
+	]
+}
+```
+
+Here we want our calls to be executed in blocks 100 (`0x64`) and in 200 (`0xc8`). The block numbers can be anything as long as they are increasing and higher than the block we are building from 10 (`0xa`). Now we end up in a situation where there exists block ranges 11-99 and 101-199 that are not defined anywhere. These blocks are called "phantom blocks". What happens if you try to request block hash of any of such blocks in the EVM? How can we calculate the block hash of future blocks when we don't know the block hash of the previous block?
+
+Our solution to this problem is to define block hash of a phantom block to be:
+
+```
+keccak(rlp([hash_of_previous_non_phantom_block, phantom_block_number]))
+```
+
+So for example in our example, you could get block hash of block 142 as follows: 
+```
+keccak(rlp([hash of block 12, 142]))
+```
+
+The phantom blocks other properties are set to their default properties as defined by the multicall specification. We came to this definition by wanting phantom block hashes to be unique if things prior to the phantom block changes, so if tooling is storing block hashes somewhere, they should remain unique if things change in the simulation.
+
+One other approach to this problem would be to really calculate the real block hashes for all the phantom blocks, but this would make generating blocks far in future really expensive, as to generate 100 phantom blocks, you would need to calculate 100 block hashes that all depend on each other. And in most cases, no one really cares about these blocks.
+
+Base fee per gas is not adjusted in the phantom blocks, their base fee remains constant.
+
+## Default values for transactions
+As multicall is an extension to `eth_call` we want to enable the nice user experience that the user does not need to provide all required values for a transaction. We are assuming following defaults if the variable is not provided by the user:
+| parameter name | description |
+-----------------|-----------------------
+| type | `0x2` |
+| nonce | Take the correct nonce for the account prior multicall and increment by one for each transaction by the account |
+| to | `null` |
+| from | `0x0000000000000000000000000000000000000000` |
+| gas limit | (blockGasLimit - SumOfGasLimitOfTransactionsWithDefinedGasLimit) / NumberOfTransactionsWithoutKnownGasLimit |
```

Comment:
Ok, this is smart! I had to play with this for a while to see if I could find an exception but obviously not, if you ensure the sum of all txs equals the block gas limit it works (ensure this is rounded down to really make sure there is no exception, so I would use `//` instead of `/` to property note this).

However there is an edge case: we can get a negative gas limit, and we can also get a gas limit which cannot cover the base gas costs of 21000 (or, if you include the calldata costs or the contract creation costs this is more). This edge case can lead to invalid blocks. (Since the txs are invalid)

The other case is the negative one, lets say we have 30M gas limit and we have 3 txs, one has gas limit 30M, the other has gas limit 30M as well (if we execute both they only use 100k gas), but then the 3rd one now has a negative gas limit of 30M which is obviously not possible.

Reply:
Ah right, I see we throw if this happens.